### PR TITLE
[threading] Implement 'pulse' transition for two-phase STW

### PR DESCRIPTION
**Pull request by lambdageek**
This addresses the symptoms (ie invalid thread state transition assertions) from #11403 and #10722, at least for GC Safe threads.

With non-hybrid suspend, a pulse is just a resume.

With hybrid suspend, a `pulse` is a combination of a resume of a preemptively
suspended GC Unsafe thread and a request that it suspend again - as if we
atomically resumed the thread and then applied the initial phase of two-phase
suspend again.  If the now-running thread comes to a GC Safe -> GC Unsafe
transition it will suspend.  If not, we will preemptively suspend it again
during the mop-up phase.

The important thing is that we don't want the thread to transition to GC Unsafe
again: if it does so, it may try to take a lock that some other
async-suspended thread is holding, and it will never reach a safepoint ---
exactly the situation that two-phase suspend is supposed to solve.

----

This doesn't fix the underlying issue in #11403 - that we sometimes see a `MonoThreadInfo*` that relates to a GCD worker thread that isn't running (because the OS reuses a `pthread_t`).  But it does mean that we will resume it without asserting and then try to suspend it again once it's running and has non-zero stack and doesn't look like it's in a critical region anymore.

It's a fix for #10722 assuming that waiting for a resume after a self-suspend doesn't end up in what looks like a WOW64 API call - that would look like a GC Unsafe thread in a critical region which we would attempt to pulse. For now `mono_threads_transition_request_pulse` will assert if it sees a GC Unsafe thread.

### CODE REVIEW

File: mono/utils/mono-threads.c [hunk 1]
```diff
@@ -1117,6 +1136,21 @@ mono_thread_info_begin_resume (MonoThreadInfo *info)
...
+mono_thread_info_begin_pulse_resume_and_request_suspension (MonoThreadInfo *info)
+{
+	/* For two-phase suspend, we want to atomically resume the thread and
+	 * request that it try to cooperatively suspend again.  Specifically,
+	 * we really don't want it to transition from GC Safe to GC Unsafe
+	 * because we then it could (in GC Unsafe) try to take a lock that's
+	 * held by another preemptively-suspended thread, essentially
+	 * recreating the same problem that two-phase suspend intends to
+	 * fix. */
+	if (!mono_threads_is_hybrid_suspension_enabled ())
```
Comment (luhenry) [inline mono/utils/mono-threads.c]: Shouldn't this be `if (! mono_threads_is_hybrid_suspension_enabled() && ! mono_threads_is_cooperative_suspension_enabled())`?

Comment (lambdageek) [reply mono/utils/mono-threads.c]: It should be a (non-existent) `mono_threads_is_multiphase_stw_enabled()`: we only need to pulse if STW requires two-phase suspension.  If there's only a single phase, resuming is fine.

**Comment by lambdageek** [comment]
Added `mono_threads_is_multiphase_stw_enabled`

**Comment by luhenry** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
@monojenkins backport to 2018-10

**Pull request merged**